Add podspec for GoogleAPIClient.

tswast opened the pull request:
This depends on having a podspec for GTMSessionFetcher https://github.com/google/gtm-session-fetcher/pull/31 and GTMOAuth2 https://github.com/google/gtm-oauth2/pull/54. Those should be merged in before this one.

We should tag a v1.0.0 release after this is pushed, too.

Please take a look. @thomasvl @jayneflux

tswast commented:
This is based loosely on the community-contributed podspec at https://github.com/CocoaPods/Specs/blob/master/Specs/Google-API-Client/1.0.422/Google-API-Client.podspec.json

perotinus commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,196 @@
+  s.ios.frameworks = 'UIKit'
+  s.dependency 'GTMOAuth2', '~> 1.1'
+  s.dependency 'GTMSessionFetcher', '~> 1.1'
+
```
I think you should have a default subspec, possible Common: I don't think people will ever want to get everything, and since there is so much, I think it's better to make users pick and choose early rather than trim later.

perotinus commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,196 @@
+  s.ios.deployment_target = '6.0'
+  s.osx.deployment_target = '10.8'
+  s.requires_arc = false
+  s.ios.frameworks = 'UIKit'
```
Does CocoaPods add this by default? I think it assumes that iOS pods require Foundation and UIKit.

tswast replied:
Seems to work without it. I'll update with a new commit soon.

perotinus commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,196 @@
+# This file specifies the Pod setup for GoogleAPIClient. It enables developers
```
Not sure you need this comment.

perotinus commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,196 @@
+  s.dependency 'GTMOAuth2', '~> 1.1'
+  s.dependency 'GTMSessionFetcher', '~> 1.1'
+
+  s.subspec 'Common' do |common|
```
In the GTM* podspecs, we named each subspec |sp| rather than having the name based on the name of the subspec. Not a big deal, but it might be good to be consistent across Google's pods.

tswast commented:
@perotinus I pushed a new commit.

I made GoogleAPIClient/Common the default subspec and named each subspec `|sp|` to be consistent with the other Google pods.

SSheldon commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,195 @@
+  s.homepage     = 'https://github.com/google/google-api-objectivec-client'
+  s.license      = { :type => 'Apache', :file => 'LICENSE' }
+  s.source       = { :git => 'https://github.com/google/google-api-objectivec-client.git'},
+                     :tag => "v#{s.version}" }
```
I'm getting an error here from `pod lib lint`:
```
[!] Invalid `GoogleAPIClient.podspec` file: syntax error, unexpected =>, expecting keyword_end
                     :tag => "v#{s.version}" }
```

Looks like there shouldn't be a closing `}` on the above line, it's choking the parser.

SSheldon commented on GoogleAPIClient.podspec:
```diff
@@ -0,0 +1,195 @@
+  s.ios.deployment_target = '6.0'
+  s.osx.deployment_target = '10.8'
+  s.requires_arc = false
+  s.dependency 'GTMOAuth2', '~> 1.1'
```
Can the `GTMOAuth2` dependency be removed? As far as I can tell, it's only ever used in the example code (which isn't included in this podspec) and I was able to compile a subspec without it. If it's not necessary, one less dependency is always nice :)

thomasvl replied:
Most of the apis end up needing auth.  So, yes it isn't always needed, but my guess is the majority case it developers will end up needing it and will likely be left trying to figure out how to get it.

SSheldon replied:
Hmm, in our use case auth is provided by the GoogleSignIn pod from [`-[GIDAuthentication fetcherAuthorizer]`](https://developers.google.com/identity/sign-in/ios/api/interface_g_i_d_authentication.html#a4bae2b2c1dacae3419014e8981b64088).

Would it be possible to allow opting out of GTMOAuth2 so that people using GoogleSignIn don't need to include it? Maybe it could be added in an `Auth` subspec that is a default subspec; then it can be opted out of by specifying specific subspecs rather than using the default.

thomasvl replied:
Under the hood there you'll likely find the same code.  Is GIDAuth going to end up pointing to the opensource specs also instead of shipping things internally?

eliotstock replied:
We also use GoogleSignIn and would prefer not to have to include GTMOAuth2 if at all possible.

SSheldon replied:
That would be nice :)

In the GoogleSignIn pod I see a lot of symbols from the open source code here but prefixed with `GSDK`, like `GSDK_GTMOAuth2Authentication` and `GSDK_GTMSessionFetcher`. If the `GoogleSignIn` pod depended on these pods we could avoid duplicating those symbols.

However I can see that potentially being a difficult change to make; then you might have a harder time distributing GoogleSignIn to people who aren't using CocoaPods.

tswast commented:
I pushed a new commit.

I renamed Common to Core. I also fixed the syntax error for the tag. I didn't know about `pod lib lint`, thanks. I was using `pod spec lint`, which didn't like the tag parameter.

thomasvl commented:
Lets go ahead and drop the gtm-oauth2 dep then, we'll document it instead.

eytanschulman commented:
@tswast Do you know at what point one might be able to access the pod?

tswast commented:
Okay. I pushed a change that drops the gtm-oauth2 dep. Sorry for late reply, I was OOO earlier this week.

eytanschulman commented:
@tswast Any idea when you will upload to CocoaPods? I'm eager to add it to my current project.

tswast commented:
@thomasvl Could you create a v1.0.0 tag?

tswast commented:
@eytanschulman Done.

eytanschulman commented:
@tswast Thank you. One question - Now that i've removed integration with an old unofficial pod and added integration of this new official one, there are a number of classes that were integral to the functionality of my little app that have been removed, such as GTMOAuth2ViewControllerTouch and GTLCalendar. Are there equivalent replacements that are outside of the pod's installation that could fix these issues?

tswast commented:
@eytanschulman For `GTMOAuth2ViewControllerTouch`, you will likely need to add the `GTMOAuth2` dependency, added in https://github.com/JakeWharton/sdk-manager-plugin/issues/101#issuecomment-169670389. We had some discussion about this, since it is not required for the build, but often needed. (An alternative is the GoogleSignIn pod https://github.com/google/google-api-objectivec-client/pull/157#discussion_r55235555 which seems to conflict with GTMOAuth2 in some way).

For GTLCalendar, you should need to depend on the [`GoogleAPIClient/Calendar`](https://github.com/google/google-api-objectivec-client/blob/a2ab7ae8aabd2d265a2fab9af5d0d88733f99ac1/GoogleAPIClient.podspec#L74) subspec. We decided not to include all the generated libraries in the default pod to keep the included code size down.

jayneflux commented:
@tswast @eytanschulman The only way to get GTMOAuth2ViewControllerTouch should be the GTMOAuth2 pod (which was being implicitly included in the previous GoogleAPIClient pod). 
GoogleSignIn and GTMOAuth2 should not conflict.  If you see conflicts, please let me know.

tswast commented:
Thanks, good to know that they shouldn't conflict.

eytanschulman commented:
@jayneflux @tswast  is GTMOAuth2 a publicly accessible podspec? I noticed the merge from google/gtm-oauth2#54 but don't see it as a subspec nor a part of the parent.

jayneflux commented:
I just checked with @perotinus and there was a lint failure when pushing that pod (related to modular headers) which he is working on. The unofficial GTMOAuth2 pod should work till that gets sorted out though.

eytanschulman commented:
@jayneflux Great, thanks for all the help, i'll try that out!

perotinus commented:
Quick update about the OAuth2 pod: I've found two issues , one around its iOS version and one around modular headers. Neither of these seems complicated to fix, but I'm going to check with @thomasvl and figure out what approach we want to take to fixing the code/spec.
